orttraining/orttraining/training_api/module.cc

```diff
@@ -2,12 +2,137 @@
 // Licensed under the MIT License.
 
 #include "core/session/inference_session.h"
+#include "core/session/environment.h"
+
 #include "orttraining/training_api/include/module.h"
+#include "orttraining/training_api/include/utils.h"
+
+using namespace onnxruntime;
 
 namespace onnxruntime {
 namespace training {
 namespace api {
 
+Status Parameter::AllocateGrad(const std::string& gradient_name, const SessionState& sess_state) {
+  // assert param is allocated
+  ORT_ENFORCE(data_.IsAllocated());
+  ORT_ENFORCE(requires_grad_);
+  gradient_name_ = gradient_name;
+  ORT_ENFORCE(OrtValueLike(sess_state, data_, gradient_).IsOK());
+  return Status::OK();
+}
+
+Status Parameter::ResetGrad() {
+  //TODO: make use of lazy_reset_grad input instead 
+  Tensor* p_tensor = gradient_.GetMutable<Tensor>();
```

Comment:
add a check for requires_grad at the begining. You can set the contract to require the parameter to be trainable parameter which this mtd is called or to return OK() if this is not a trainable parameter.